### PR TITLE
GH-44276: [C++] Fix UBSAN error introduced by Decimal32

### PR DESCRIPTION
**Pull request by zeroshade**
### Rationale for this change
Fixing UBSAN errors caused by Decimal32/Decimal64 addition

### What changes are included in this PR?
Wrapping left shifts in `SafeLeftShift` to avoid UBSAN errors.

* GitHub Issue: #44276

### CODE REVIEW

**Review by pitrou — APPROVED** [review]
LGTM, let's just wait for CI

**Comment by kou** [comment]
@zeroshade Could you use our merge script https://github.com/apache/arrow/blob/main/dev/merge_arrow_pr.py instead of GitHub merge button next time?
It not only merge a PR but also close the associated issue and set milestone to the associated issue.

**Pull request merged**
